
---
title: |
  ![](img/CoVaRR-Net_Logo-600.png){width=1lm}  
subtitle: "Genomic epidemiology analyses and mathematical modelling notebook"
author: "Pillar 6"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
---

```{r setup, include=FALSE, warning=FALSE}

#coding and data
library(tidyverse) # wrangling and everything really
library(knitr) # Needed to set root directory

library(lubridate) # dates are special

#phylo-specific
library(treeio)  # https://bioconductor.org/packages/release/bioc/html/treeio.html
library(phylotools)
library(tidytree)
#library(phangorn)
library(phytools)

#plotting and tables
library(ggplot2) # Work-horse plotting package
library(ggtree) #All things phylogenetic visualization

#colors
library(RColorBrewer)
library(colorspace)
library(viridis)

library(MASS)

library(r2d3, quietly=TRUE)

theme_set(theme_classic())

# You would need to change this folder to be wherever you wanted the html file to be written.
opts_knit$set(root.dir = getwd())
```


```{r}

VirusSeq_date="2022-04-28"


# this can be made more compact for faster loading
meta <- read.csv(gzfile("data_needed/virusseq.metadata.csv.gz"))
meta$sample.collection.date <- as.Date(meta$sample.collection.date)
meta$province <- meta$geo_loc_name..state.province.territory.


# FIXME: restore pangolearn calls overriden by scorpio - issue #46
plearn <- gsub(".+assignment ([A-Z]+\\.[0-9.]*).*", "\\1",
               meta$note[meta$lineage=="None"])
plearn[which(grepl("_", plearn))] <- "None"
meta$lineage[meta$lineage=="None"] <- plearn

#make a pango.group column
VOCVOI <- data.frame(
  name=c('Alpha', 'Beta', 'Gamma', 'Delta', 'Delta AY.25', 'Delta AY.27', 
         'Lambda', 'Omicron BA.1', 'Omicron BA.1.1', 'Omicron BA.2', 'Mu',
         'A.23.1', 'B.1.438.1'),
  pattern=c('^B\\.1\\.1\\.7|^Q\\.', '^B\\.1\\.351', '^P\\.', 
            '^B\\.1\\.617|^(?!AY\\.2[57])AY\\.', '^AY\\.25', '^AY\\.27', 
            '^C\\.37|^C\\.37\\.1',
            '^B\\.1\\.1\\.529|^BA\\.1$|^BA\\.1\\.[1-9][0-9]|BA\\.1\\.[2-9]$', 
            '^BA\\.1\\.1$|^BA\\.1\\.1\\.[0-9]+$', 
            '^BA\\.2', 
            '^B\\.1\\.621', 
            '^A\\.23\\.1', '^B\\.1\\.438\\.1'),
  color=c('#B29C71', '#F08C3A', '#444444', '#A6CEE3', '#61A6A0',
          '#438FC0', '#CD950C', '#8B0000', '#FA8072', '#FF0000',
          '#BB4513', '#9AD378', '#3EA534')
)
variants <- sapply(VOCVOI$pattern, function(p) 
  grepl(p, meta$lineage, perl=T))

# verify that every row matches either 0 or 1 patterns
# table(apply(variants, 1, sum))

meta$pango.group <- 'other'
meta$pango.group[apply(variants, 1, sum)==1] <- VOCVOI$name[unlist(apply(variants, 1, which))]
meta$pango.group <- as.factor(meta$pango.group)

pal <- VOCVOI$color
names(pal) <- VOCVOI$name
pal["other"] <- 'grey'  # named character vector

## 2. LOAD epidemiological data (PHAC)


#from: https://health-infobase.canada.ca/covid-19/epidemiological-summary-covid-19-cases.html?stat=num&measure=total&map=pt#a2
#epidataCANall <- read.csv(url("https://health-infobase.canada.ca/src/data/covidLive/covid19-download.csv"))
#epidataCANall$date <- as.Date(epidataCANall$date)
#epidataCANall$prname <- gsub('_', ' ', epidataCANall$prname)
#epidate <- tail(epidataCANall,1)$date #download date

# for barplots
meta$week <- cut(meta$sample.collection.date, 'week')
```


```{r}
meta1 <- meta[as.Date(meta$week) > as.Date('2020-11-01'), ]
meta1$week <- as.factor(as.character(meta1$week))
tab <- as.data.frame.matrix(table(meta1$week, meta1$pango.group))
tab$week <- row.names(tab)

r2d3(data=tab, script="js/barplot.js", container="div", elementId="barplot-element")
```


