

<u id='JN_1'>JN.1</u> and its subvariants now predominate. Saltations (a variant with a sudden large increase in number of mutations, like the recently detected <u id='BA_2_87_1'>BA.2.87.1</u>) are being closely tracked, plus <u id='JN_1'>JN.1</u> subvariants with additional proposed ACE2 receptor binding enhancing mutations and immune evasive mutations (such as <u id='JN_1_4'>JN.1.4</u> and <u id='JN_1_7'>JN.1.7</u> with both S:T572I and S:R346T together).



**Variants of current interest:**

(Due to their current/potential growth advantage, mutations of potential functional significance, or spread in other countries.)



* <u id='BA_2_87_1'>BA.2.87.1</u> (saltation with a large number of spike mutations)
* <u id='JN_1'>JN.1</u> (<u id='BA_2_86_1'>BA.2.86.1</u> with immune evasive mutation S:455S - still showing notable growth in Canada).
* <u id='JN_1_4'>JN.1.4</u> (<u id='JN_1'>JN.1</u> with ORF1a:T170I, the most growing variant with a significant number of cases, other than <u id='JN_1'>JN.1</u>, with a particular interest in those with both S:T572I and S:R346T together)
* <u id='JN_1_7'>JN.1.7</u> (<u id='JN_1'>JN.1</u> with S:T572I and S:E1150D, or others with S:T572I like <u id='JN_1_8_1'>JN.1.8.1</u>)
* <u id='JN_1_6_1'>JN.1.6.1</u> (<u id='JN_1_6'>JN.1.6</u> with S:R346T, and other variants with immune evasive mutations like S:R346T.

â€¦plus any highly divergent variants and sublineages with additional combinations of the mutations below.



* NSP13:V348T (also known as ORF1b:V1273T - a rare two-nucleotide mutation that is TRS-related. See the excellent summary by Ryan Hisner at <https://twitter.com/LongDesertTrain/status/1745578135033110737>)
* S:683W and other furin cleavage site mutations that may impact/reduce disease severity.
* S:572I which is thought to increase ACE2 receptor binding.

Plus other mutations identified through mutation scanning that are involved in binding or immune evasion (like S:346T). See:



* Greaney, Starr, &amp; Bloom, Virus Evolution, 8:veac021 (2022)
* Cao et al, Nature, 614:521-529 (2023)
* Yisimayi et al, bioRxiv, DOI 10.1101/2023.05.01.538516 (2023)
* Dadonaite et al, bioRxiv, DOI 10.1101/2023.11.13.566961 (2023)
* Bdeir et al, medRxiv, DOI 10.1101/2024.01.03.23300575 (2024)

With thanks to the global team of variant hunters, and other SARS-CoV-2 genome analysis tool providers (see List of Useful Tools below), which play a key role in identifying new variants of note. (edited)


<!-- edited -->