

<u id='JN_1'>JN.1</u> (<u id='BA_2_86_1'>BA.2.86.1</u> with immune evasive mutation S:455S) has continued to grow in proportion of variants. <u id='JN_1'>JN.1</u> subvariants, including those with mutations that likely impact ACE2 binding or immune evasion, are being closely tracked. <u id='JN_2_5'>JN.2.5</u>, which independently obtained S:455S, is also showing notable growth, but primarily in those regions with less <u id='JN_1'>JN.1</u>.



**Variants of current interest:**

(Due to their current/potential growth advantage, mutations of potential functional significance, or spread in other countries.)



* <u id='JN_1'>JN.1</u> (<u id='BA_2_86_1'>BA.2.86.1</u> with immune evasive mutation S:455S - still showing notable growth in Canada) plus many subvariants which are showing similar growth including those below.
* <u id='JN_1_4'>JN.1.4</u> (<u id='JN_1'>JN.1</u> with ORF1a:T170I)
* <u id='JN_1_5'>JN.1.5</u> (<u id='JN_1'>JN.1</u> with NSP13:V348T, also known as ORF1b:V1273T, with just a few cases in Canada to date, but notable growth in some regions internationally).
* <u id='JN_1_6_1'>JN.1.6.1</u> (<u id='JN_1_6'>JN.1.6</u> with S:R346T, and other variants with immune evasive mutations like S:R346T)
* <u id='JN_1_7'>JN.1.7</u> (<u id='JN_1'>JN.1</u> with S:T572I and S:E1150D, or others with S:T572I)

â€¦plus any highly divergent variants and sublineages with additional combinations of the mutations below.



* NSP13:V348T (also known as ORF1b:V1273T - a rare two-nucleotide mutation that is TRS-related. See the excellent summary by Ryan Hisner at <https://twitter.com/LongDesertTrain/status/1745578135033110737>)
* S:683W and other furin cleavage site mutations that may impact/reduce disease severity.
* S:572I which is thought to increase ACE2 receptor binding.

Plus other mutations identified through mutation scanning that are involved in binding or immune evasion (like S:346T). See:



* Greaney, Starr, & Bloom, Virus Evolution, 8:veac021 (2022)
* Cao et al, Nature, 614:521-529 (2023)
* Yisimayi et al, bioRxiv, DOI 10.1101/2023.05.01.538516 (2023)
* Dadonaite et al, bioRxiv, DOI 10.1101/2023.11.13.566961 (2023)
* Bdeir et al, medRxiv, DOI 10.1101/2024.01.03.23300575 (2024)

With thanks to the global team of variant hunters, and other SARS-CoV-2 genome analysis tool providers (see List of Useful Tools below), which play a key role in identifying new variants of note.


<!-- edited -->