

JN.1 has continued to grow as a proportion of variants. JN.1 subvariants are being closely tracked, including those with additional proposed ACE2 receptor binding enhancing mutations like S:572I (obtained independently multiple times) and immune evasive mutations.


**Variants of current interest:**

(Due to their current/potential growth advantage, mutations of potential functional significance, or spread in other countries.)


* JN.1 (BA.2.86.1 with immune evasive mutation S:455S - still showing notable growth in Canada).
* JN.1.7 (JN.1 with S:T572I and S:E1150D, or others with S:T572I like JN.1.8.1)
* JN.1.6.1 (JN.1.6 with S:R346T, and other variants with immune evasive mutations like S:R346T, including recent occurrences of both S:T572I and S:R346T together)

â€¦plus any highly divergent variants and sublineages with additional combinations of the mutations below.



* NSP13:V348T (also known as ORF1b:V1273T - a rare two-nucleotide mutation that is TRS-related. See the excellent summary by Ryan Hisner at https://twitter.com/LongDesertTrain/status/1745578135033110737)
* S:683W and other furin cleavage site mutations that may impact/reduce disease severity.
* S:572I which is thought to increase ACE2 receptor binding.


Plus other mutations identified through mutation scanning that are involved in binding or immune evasion (like S:346T). See:



* Greaney, Starr, & Bloom, Virus Evolution, 8:veac021 (2022)
* Cao et al, Nature, 614:521-529 (2023)
* Yisimayi et al, bioRxiv, DOI 10.1101/2023.05.01.538516 (2023)
* Dadonaite et al, bioRxiv, DOI 10.1101/2023.11.13.566961 (2023)
* Bdeir et al, medRxiv, DOI 10.1101/2024.01.03.23300575 (2024)

With thanks to the global team of variant hunters, and other SARS-CoV-2 genome analysis tool providers (see List of Useful Tools below), which play a key role in identifying new variants of note.


<!-- edited -->