---
title: "CoVaRR-Net"
subtitle: "Genomic epidemiology analyses and mathematical modelling notebook"
author: "Pillar 6"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
---

```{r setup, include=FALSE, warning=FALSE}

#coding and data
library(tidyverse) # wrangling and everything really
library(knitr) # Needed to set root directory
library(reticulate) # Needed for the Python code chunk 
#py_install("pandas") #install pandas in virtual environment
#use_virtualenv(virtualenv = "r-reticulate")
use_condaenv("r-reticulate")
#knitr::knit_engines$set(python = reticulate::eng_python)
library(lubridate) # dates are special
#phylo-specific
library(treeio)
library(phylotools)
library(tidytree)
library(phangorn)
library(phytools)
#plotting and tables
library(ggplot2) # Work-horse plotting package
library(ggtree) #All things phylogenetic visualization
library(cowplot) # Needed to make easy multi-panel plots with ggplot2
library(DT) # Needed to make DT example table
library(ggbeeswarm) # Needed to make beeswarm plots
library(gridExtra) # multi-panel plots
library(kableExtra) # Needed to write out a formatted table
library(scales)
#colors
library(RColorBrewer)
library(colorspace)
library(viridis)
library(MASS)

#library(bbmle)
#library(HelpersMG)

theme_set(theme_classic())

# # This is the Python to be used (which you would only specify if using Python code chunks)
# use_python("/Users/carme/local/miniconda3/bin/python") # <- FIX this path
if(strsplit(getwd(), "/")[[1]][3]=="rp"){#python path that works for Raphael
  Sys.setenv(RETICULATE_PYTHON = "/media/rp/partition4data/Dropbox/Raph/covarrnet/python_R_env/bin/python")
}else{#python path that works for Susanne
   Sys.setenv(RETICULATE_PYTHON = "my_env/bin/python")
}
# You would need to change this folder to be wherever you wanted the html file to be written.
opts_knit$set(root.dir = getwd())
```

### Contributing authors:  

Data analysis, code, and maintenance of this notebook: Carmen Lia Murall, RaphaÃ«l Poujol, Susanne Kraemer, Arnaud N'Guessan, Sarah Otto, Art Poon, Jesse Shapiro. Input and direction by other members of Pillar 6 (https://covarrnet.ca/our-team/#pillar-6 ), which include: Fiona Brinkman, Caroline Colijn, Jorg Fritz, Morgan Langille, Paul Gordon, Julie Hussin, Jeff Joy, William Hsiao, and Erin Gill. 

Sequence collection, generation, release, and feedback on analyses: Canadian laboratories as part of the CPHLN and CanCOGeN are making these data publicly available and contribute feedback on analyses presented here. A complete list of lab authors is in this repository, and more details are below in the Acknowledgement section.   


# Introduction  

This notebook is built to explore Canadian SARS-CoV-2 genomic and epidemiological data with the aim of investigating viral evolution and spread. It is for discussion with pillar 6's team and for sharing with collaborators, e.g. PH labs. These analyses can spur further research within or across pillars, be used for reports (or data dashboards), support discussions with the science communication pillar for public dissemination, and enable code reuse by public health authorities/laboratories for their internal use.

Canadian genomic and epidemiological data will be regularly pulled from various public sources (see list below) to keep these analyses up-to-date. Only representations of aggregate data will be posted here.


```{r load_data}

## 1. LOAD processed metadata of Canadian sequences (with latest pangolin, division, and full seq IDs)
#Download metadata from gisaid, put the date here:

gisaiddate="2022_03_21"


# this can be made more compact for faster loading
meta <- read.csv(gzfile("data_needed/virusseq.2022-03-16T15:17:45.csv.gz"))
meta$sample.collection.date <- as.Date(meta$sample.collection.date)
meta$province <- meta$geo_loc_name..state.province.territory.


#make a pango.group column
VOCVOI <- data.frame(
  name=c('Alpha', 'Beta', 'Gamma', 'Delta', 'Delta AY.25', 'Delta AY.27', 
         'Lambda', 'Omicron BA.1', 'Omicron BA.1.1', 'Omicron BA.2', 'Mu',
         'A.23.1', 'B.1.438.1'),
  pattern=c('^B\\.1\\.1\\.7|^Q\\.', '^B\\.1\\.351', '^P\\.', 
            '^B\\.1\\.617|^(?!AY\\.2[57])AY\\.', '^AY\\.25', '^AY\\.27', 
            '^C\\.37|^C\\.37\\.1', '^B\\.1\\.1\\.529|^BA\\.1$', 
            '^BA\\.1\\.1', '^BA\\.2', '^B\\.1\\.621', 
            '^A\\.23\\.1', '^B\\.1\\.438\\.1'),
  color=c('#B29C71', '#F08C3A', '#444444', '#A6CEE3', '#61A6A0',
          '#438FC0', '#CD950C', '#8B0000', '#FA8072', '#FF0000',
          '#BB4513', '#9AD378', '#3EA534')
)
variants <- sapply(VOCVOI$pattern, function(p) 
  grepl(p, meta$lineage, perl=T))

# verify that every row matches either 0 or 1 patterns
# table(apply(variants, 1, sum))

meta$pango.group <- 'other'
meta$pango.group[apply(variants, 1, sum)==1] <- VOCVOI$name[unlist(apply(variants, 1, which))]
meta$pango.group <- as.factor(meta$pango.group)

pal <- VOCVOI$color
names(pal) <- VOCVOI$name
pal["other"] <- 'grey'  # named character vector

## 2. LOAD epidemiological data (PHAC)


#from: https://health-infobase.canada.ca/covid-19/epidemiological-summary-covid-19-cases.html?stat=num&measure=total&map=pt#a2
epidataCANall <- read.csv(url("https://health-infobase.canada.ca/src/data/covidLive/covid19-download.csv"))
epidataCANall$date <- as.Date(epidataCANall$date)
epidataCANall$prname <- gsub('_', ' ', epidataCANall$prname)
epidate <- tail(epidataCANall,1)$date #download date

# for barplots
require(lubridate)
meta$week <- cut(meta$sample.collection.date, 'week')
meta1 <- meta[as.Date(meta$week) > as.Date('2020-11-01'), ]
meta1$week <- as.factor(as.character(meta1$week))
```

Mutational profile of Omicron and its sublineages in Canada and globally for the most prevalent (>75%) point mutations in each category (based on all genomes available on GISAID on `r gisaiddate`.




<!-- # Evolution and growth of SARS-CoV-2 in Canada   -->

<!-- There are various methods to investigate changes in evolutionary rates of VOC/VOIs and to compare their relative fitness in an epidemiological context. Here we present two of such methods. -->


## Selection on omicron sublineages {.tabset}  

Of particular interest when there are newly arriving or emerging lineages is whether they have a selective advantage (and by how much), relative to the lineages already circulating within a population. Here we examine the major sublineages currently in Canada and their relative rate of spread. 

Currently, the major variants circulating are Omicron sublineages BA.1, BA.1.1, and BA.2. Sub-lineage BA.1 was initially prevalent, but Omicron sublineages BA.1.1 and BA.2 have been spreading, as illustrated in the plots below. Left plot: y-axis is the proportion of sublineages BA.1.1 and BA.2 among Omicron; right plot: y-axis describes the logit function, log(freq(BA.1.1 or BA.2)/freq(BA.1)), which gives a straight line whose slope is the selection coefficient if selection is constant over time (see methods). 

For comparison, Alpha had a selective advantage of 6-11% per day over preexisting
SARS-CoV-2 lineages, and Delta had a selective advantage of about 10% per day over Alpha.

**Caveat**
Selection coefficients are not estimated for Alberta, which is currently taking a variant-specific sequencing strategy, based on an initial PCR screen, which would skew estimates of selection. Canada-wide estimates, thus, do not include this province. Separate analyses are provided for those provinces with at least 20 BA.2 sequences in the database

```{r setting function for selection estimator}
name1<-"BA.1" #Can include a list here (first value is used as plot labels)
name2<-"BA.1.1" #Can include a list here (first value is used as plot labels)
name3<-"BA.2" #Can include a list here (first value is used as plot labels)

#color for each lineage
col2=pal[paste0("Omicron ",name2)]
col3=pal[paste0("Omicron ",name3)]

#Set a starting date
#Note that the startdate shouldn't be too much before both alleles become common
#or rare migration events that die off could throw off the estimation procedure 
#(so that the parameter estimates account for the presence of those alleles long in the past).
startdate<-as.Date("2021-12-15") #Using a later date with less sampling noise
source("scripts/plot_selection_estimator3.R")
source("scripts/plot_selection_estimator2.R")
#####################################################
# tabs for displaying in notebook
#each PT tab should have curve plot and breakpoint plot side by side
```


### Canada
#### Canada without Alberta

```{r selection estimator Canada, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator3("Canada (no AB)",startdate,name1,name2,name3,col2,col3)
plot_selection_estimator3("Saskatchewan",startdate,name1,name2,name3,col2,col3)
```

### BC
#### British Columbia

```{r selection estimator BC, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
mydata=meta %>% filter(grepl("BA.", lineage), geo_loc_name..state.province.territory. == "Saskatchewan", !is.na(sample.collection.date), sample.collection.date >= startdate) %>% group_by(sample.collection.date) %>% count(lineage)



print(mydata)
lastdate<-max(mydata$sample.collection.date)

#convert time to an integer counter for use in fitting, first using the last date as time 0:
mydata$time = as.numeric(difftime(mydata$sample.collection.date, lastdate, units = "days"))

#filter data to after that starting date
data1 <- filter(mydata, lineage %in% name1)
data2 <- filter(mydata, lineage %in% name2)
data3 <- filter(mydata, lineage %in% name3)

#allow multiple Pango lineages to be combined if name1 or name2 includes a list, summing n
data1 <- as.data.frame(unique(data1 %>% group_by(time) %>% transmute(day=sample.collection.date,  n=sum(n), time=time)))
data2 <- as.data.frame(unique(data2 %>% group_by(time) %>% transmute(day=sample.collection.date,  n=sum(n), time=time)))
data3 <- as.data.frame(unique(data3 %>% group_by(time) %>% transmute(day=sample.collection.date,  n=sum(n), time=time)))
name1 <- name1[[1]]
name2 <- name2[[1]]
name3 <- name3[[1]]
data1$lineage <- name1
data2$lineage <- name2
data3$lineage <- name3

#join lists in a dataframe to plot proportions and represent time as a list of integers
timestart<-as.numeric(difftime(startdate, lastdate, units = "days"))
timeend<-as.numeric(difftime(lastdate, lastdate, units = "days"))
toplot <- data.frame(time = seq.int(timestart,timeend))
toplot$n1 <- data1$n[match(toplot$time,data1$time)]
toplot$n2 <- data2$n[match(toplot$time,data2$time)]
toplot$n3 <- data3$n[match(toplot$time,data3$time)]
toplot[is.na(toplot)] = 0 #Any NA's refer to no variant of that type on a day, set to zero


#To aid in the ML search, we rescale time to be centered as close as possible
#to the midpoint for the second variable (p=0.5), to make sure that the alleles 
#are segregating at the reference date.
#If we set t=0 when p is near 0 or 1, then the likelihood surface is very flat.
v=toplot$n1*toplot$n2*toplot$n3
refdate<-which(v==max(v,na.rm=TRUE))
refdate<-refdate[[1]] #Just in case there is more than one matching point, the first is taken
timeend <- (timeend-timestart)-refdate
timestart <- -refdate
toplot$time <- seq.int(timestart,timeend)
data1$time <- data1$time + (timeend-timestart)-refdate
data2$time <- data2$time + (timeend-timestart)-refdate
data3$time <- data3$time + (timeend-timestart)-refdate

#date converter
dateseq <- seq.Date(startdate,lastdate,"days")
dateconverter <- data.frame(time=toplot$time,date=dateseq)

startp <- 0.5

trifunc <- function(p2,p3,s2,s3){
  print(paste("p2",str(p2)))
  print(paste("p3",str(p3)))
  print(paste("s2",str(s2)))
  print(paste("s3",str(s3)))
  -(sum(data1$n*log((1-p2-p3)/((1-p2-p3)+p2*exp(s2*data1$time)+p3*exp(s3*data1$time))))+
      sum(data2$n*log(p2*exp(s2*data2$time)/((1-p2-p3)+p2*exp(s2*data2$time)+p3*exp(s3*data2$time))))+
      sum(data3$n*log(p3*exp(s3*data3$time)/((1-p2-p3)+p2*exp(s2*data3$time)+p3*exp(s3*data3$time)))))
}

p2=1.4
p3=0.0499
s2=0.0998
s3=0.538

print((1-p2-p3))
print((1-p2-p3)/((1-p2-p3)+p2*exp(s2*data1$time)+p3*exp(s3*data1$time)))
print(sum(data1$n*log((1-p2-p3)/((1-p2-p3)+p2*exp(s2*data1$time)+p3*exp(s3*data1$time)))))
print(sum(data2$n*log(p2*exp(s2*data2$time)/((1-p2-p3)+p2*exp(s2*data2$time)+p3*exp(s3*data2$time)))))
print(sum(data3$n*log(p3*exp(s3*data3$time)/((1-p2-p3)+p2*exp(s2*data3$time)+p3*exp(s3*data3$time)))))
print(log(p3*exp(s3*data3$time)/((1-p2-p3)+p2*exp(s2*data3$time)+p3*exp(s3*data3$time))))
print(p3*exp(s3*data3$time)/((1-p2-p3)+p2*exp(s2*data3$time)+p3*exp(s3*data3$time)))

bbml<-mle2(trifunc, start = startpar, lower=c(p2=10^-5, p3=10^-5),upper=c(p2=1-10^-5,p3=1-10^-5))

print(data2)

bbml

#lnL
bbml.value<--bbml@min


plot_selection_estimator3("British Columbia",startdate,name1,name2,name3,col2,col3)

```

### SK
#### Saskatchewan

```{r selection estimator SK, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator3("Saskatchewan",startdate,name1,name2,name3,col2,col3)
```


### MB
#### Manitoba

```{r selection estimator MB, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator2("Manitoba",startdate,name1,name2,col2)
```

### ON
#### Ontario

```{r selection estimator ON, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator3("Ontario",startdate,name1,name2,name3,col2,col3)
```


### ON2
#### Ontario

```{r selection estimator ON2, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator2("Ontario",startdate,name1,name2,col2)
```

### QC
#### Quebec

```{r selection estimator QC, message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator2("Quebec",startdate,name1,name2,col2)
```

### East
#### Quebec

```{r selection estimator East prov., message=FALSE , warning=FALSE, echo = FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
plot_selection_estimator2("East provinces (NL+NS+NB)",startdate,name1,name2,col2)
```

## Show

```{r session_info}
sessionInfo()
```
