```{r, warning=FALSE, message=FALSE, echo=FALSE}
source("scripts/plot_selection_estimator.R")

#define starting parameters (p,s) for making the selection estimates
#startpar <- list(p=c(0.2, 0.05), s=c(0.05, 0.1))
startpar <- list(p=c(0.1, 0.05, 0.01), s=c(0.05, 0.05, 0.01))
  
#filter samples with variants of interest
setAll=getAllStrictoLineages(meta) 
#sublineages_BQ <- getStrictoSubLineages("BQ*",meta) #BQ variants
sublineages_XBB <- getStrictoSubLineages("XBB*",meta) #XBB variants

sublineages_XBB15  <- getStrictoSubLineages("XBB.1.5*",meta) #XBB1.5 variants
sublineages_XBB116 <- getStrictoSubLineages("XBB.1.16*",meta)
sublineages_XBBOther <- sublineages_XBB[!sublineages_XBB %in% c(sublineages_XBB15,sublineages_XBB116)] #XBB Other

#setAll=setAll[!setAll %in% c(sublineages_BQ, sublineages_XBB)] #the rest
setAll=setAll[!setAll %in% c(sublineages_XBB116, sublineages_XBB15, sublineages_XBBOther)] #the rest

#define the mutants
#mutants = list(sublineages_BQ, sublineages_XBB )
#mutantNames = list("BQ*", "XBB*", "the rest") #
#col <- c(pal["Omicron BQ"], "XBB*"=pal["XBB"]) #define custom color for XBB here because otherwise it will be black cuz recombinant.

mutants = list(sublineages_XBB15, sublineages_XBB116, sublineages_XBBOther )
#this list defines the mutants we are looking at for plotting. The order of mutants listed will be respected on the respective plots. 
#!!!"The Rest" should always be last.
mutantNames = list("XBB.1.5*", "XBB.1.16*", "XBB.Other", "The Rest") #
col <- c("XBB.1.5*"="#CC5500", "XBB.1.16*" = "#00eeff", "XBB.Other"=pal["XBB"]) #define custom color for XBB here because otherwise it will be black cuz recombinant.
#im cheating, using zero width spaces to sort text. dont hate me.

#Set a starting date
#Note that the startdate shouldn't be too much before both alleles become common
#or rare migration events that die off could throw off the estimation procedure 
#(so that the parameter estimates account for the presence of those alleles long in the past).
startdate<-max(meta$sample_collection_date)-days(120) #Using a later date with less sampling noise
maxdate<-params$datestamp

#wrapper for calling plot.selection.estimate.ggplot() to avoid passing in the same local variables multiple time anjd to include a trycatch statement to show empty plots.
sub.plot.selection.estimate <- function(region,maxdate=NA){
  #plot.selection.estimate.ggplot() returns a named list of data used for plotting the case count selection curves, see the function documentation for details. 
  obj <- tryCatch(
    {
        obj=plot.selection.estimate.ggplot(region=region,
                                        startdate=startdate, startpar=startpar,
                                        reference=c(setAll),
                                        mutants=mutants,
                                        names=mutantNames,
                                        maxdate=maxdate, col=col)
    },
    error=function(cond){
        #print(cond)
        obj= (list("plot1"=NULL, "plot2"=cond))
    })
  return(obj)
}

#calculate estimates for each province
selectionEstimateFits <- list() #named list to store estimates for different provinces
selectionEstimateFits[["Canada"]] = sub.plot.selection.estimate(region="Canada", maxdate)
#maxdate = selectionEstimateFits[["Canada"]]$date

for (i in 2:length(all.regions$name)){
  selectionEstimateFits[[all.regions$shortname[[i]]]] = sub.plot.selection.estimate(region=all.regions$name[[i]], maxdate)
}


```

# Selection on Omicron

## {.tabset .tabset-fade .unlisted .unnumbered .toc-ignore}

Here we examine the relative rate of spread of the different sublineages of Omicron currently in Canada. Specifically, we determine if a new or emerging lineage has a selective advantage (s), and by how much, against a reference lineage previously common in Canada (see the methods for more details about selection and how it is estimated).

Currently, the major group of Omicron lineages rising in frequency are the XBB.* group, dominated by the XBB.1.5* lineages which is near it's peak. There are also several other XBB lineages showing a selective advantage. The previous dominant lineage, BQ.*, is nearly at the bottom, however several of it's descendant are beginning to grow again. See Fastest Growing Lineages for details on lineages showing positive selection.

We first show this growth of XBB.1.5\*, XBB.1.16\*, or XBB.Other, relative to the remaining strains, which consist predominantly of BQ.\*. Left plot: y-axis is the proportion of sub-lineages XBB.1.5\*, XBB.1.16\*, and XBB.Other relative to the remaining strains; right plot: y-axis describes the logit function, log(freq(XBB.1.5\*, XBB.1.16\*, or XBB.Other)/freq(the rest)), which gives a straight line whose slope is the selection coefficient if selection is constant over time (see methods).

For comparison, Alpha had a selective advantage of s ~ 6%-11% per day over preexisting SARS-CoV-2 lineages, and Delta had a selective advantage of about 10% per day over Alpha.

**Caveat:** These selection analyses must be interpreted with caution due to the potential for non-representative sampling, lags in reporting, and spatial heterogeneity in prevalence of different sublineages across Canada. Provinces that do not have at least 20 sequences of a lineage during this time frame are not displayed.

```{r, warning=FALSE, message=FALSE, fig.height=8, fig.width5, warning=FALSE, out.width="50%", results='asis', echo=FALSE}
#define provinces to show selection plots for.
dataAvailable.regions <- all.regions %>% filter(shortname %in% c("Canada", "BC", "AB", "SK", "MA", "ON", "QC", "East"))
#display all selection plots with each province as a tab.
apply(all.regions,1,function(reg){
  cat("###", reg[["shortname"]], " {.unlisted .unnumbered}\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("####",reg[["name"]],"\n","\n")
  if (is.null(selectionEstimateFits[[reg[["shortname"]]]]$plot1))
  {print(getEmptyErrorPlotWithMessage("Not enough Data Available"))
    print(getEmptyErrorPlotWithMessage("Not enough Data Available"))}
  else {
    print(selectionEstimateFits[[reg[["shortname"]]]]$plot1)
    print(selectionEstimateFits[[reg[["shortname"]]]]$plot2)
  }
  cat("\n\n")})
  
```

# Case count trends by variant

## {.tabset .tabset-fade .unlisted .unnumbered .toc-ignore}

These plots follow reported cases per 100,000 individuals (green dots), ignoring the most recent case counts (hollow circles), which are generally underestimated as data continues to be gathered. A cubic spline is fit to the log of these case counts to illustrate trends (top curve). The last two days of accurate case counts are then used to estimate the daily exponential growth rate $r$ in COVID-19 cases. The fit from the "Selection on Omicron" section above is used to show how each of the sub-lineages is growing or shrinking, with the corresponding growth rate $r$ for each sub-lineage on the last two days of accurate case counts. For detailed methodology, please see the methods section in the [appendix](#Methodology).

```{r case count selection estimator setup, echo=FALSE, message=F, warning=F}
source("scripts/plot_case_count_selection_estimator.R")

#loads in the case count data for each province into a named list.
caseData <- parseCaseData(all.regions, params$datestamp, params$datadir)
#loads in population data for /100000invidivudal normalization
populationData <- read.table("resources/2022Population.tsv", header=T, sep='\t', check.names = F)
#view(caseCountData)
#' function used to generate data required case count selection plots.
#' @param caseCountData The case counts for a specific province. i.e. one item in variable caseData
#' @param selectionEstimateObject The selection estimate object for a specific province. i.e. one item in variable selectionEstimateFits
#' @param filename String. If not NA, it will save a local copy of a plot to $PWD
plotCaseCountSelection <- function(caseCountData, selectionEstimateObject, saveToFile=F){
  #caseCountData <- caseData$QC
  #selectionEstimateObject<- selectionEstimateFits$QC
  #filename=NA
  if (selectionEstimateObject$region == "Canada"){
      dayToCut = 1
      #caseCountData <- head(caseCountData, -1)
    } else { 
      dayToCut = 1 
    } #define the number of days since most recent data date as "underreported"
  #populate a column with reporting accuracy (Accurate | UnderReported])
  caseCountData$report_type <- "Accurate"
  caseCountData$report_type[(nrow(caseCountData)-dayToCut+1):nrow(caseCountData)] <- "UnderReported"
  #construct a spline fit for the accurate case counts only
  caseFitModel <- getCaseCountSmoothFitWithLambda(caseCountData %>% filter(report_type=="Accurate"))
  #extract the y-values for the fit
  caseFitLineValues <- 10^(caseFitModel$y)
  #extend the previous fit for the "underreported dates" and extract the y values
  caseFitLinePredictedValues <- c((10^(predict(caseFitModel, data.frame(x=(nrow(caseCountData)-dayToCut+1):nrow(caseCountData)))[[2]]))$x)
  #merge the accurate and underreported case count fit lines and construct a DF with X, and metadata
  caseCountFitLine<-data.frame(caseCountData$Reported_Date, c(caseFitLineValues, c(caseFitLinePredictedValues)), caseCountData$report_type)
  colnames(caseCountFitLine) <- c("Reported_Date", "n", "type")
  #replaces UnderReported with Projected in the "type" column
  caseCountFitLine <- caseCountFitLine %>% mutate(type = ifelse(type=="UnderReported", "Projected", "Actual"))

  #get the selection estimate fits, scurves and extended scurves from the selection estimate object
  selectionFit <- selectionEstimateObject$fit
  selectionScurves <- selectionEstimateObject$scurves
  selectionScurvesExtended <- selectionEstimateObject$scurvesExtended
  #since canada only have data once every 7 days, modify the estimate fit to reflect that.
  if (!selectionEstimateObject$region %in% c("Alberta", "Ontario", "Quebec")){
      selectionScurves <- selectionEstimateObject$scurves[seq(1, nrow(selectionEstimateObject$scurves), 7), ]
      selectionScurvesExtended <- selectionEstimateObject$scurvesExtended[seq(1, nrow(selectionEstimateObject$scurvesExtended), 7), ]
  }
  #assign names and colors 
  selectionNames <- selectionEstimateObject$names
  selectionColors <- selectionEstimateObject$color
  
  #now we beging constructing the object that contains data for different lines to plot.
  caseSelectionLines <- list()
  #build to total case line.
  caseSelectionLines[[1]] <- list("line"=caseCountFitLine, "names" = "CaseCount", "color"="limegreen")

  #build line for each variant
  for (i in seq(from=2, to=ncol(selectionScurves))){
      x = caseCountData$Reported_Date
      actual = caseCountFitLine$n[1:length(selectionScurves[,i])] * selectionScurves[,i]
      actual <- actual[!is.na(actual)]
      if (length(actual) == length(caseCountData$Reported_Date)){
        proj = NULL
      } else{
        proj = caseCountFitLine$n[(length(selectionScurves[,i])+1):length(caseCountFitLine$n)] *   selectionScurvesExtended[,i][(length(selectionScurves[,i])+1):length(caseCountFitLine$n)]
        proj <- proj[!is.na(proj)]
      }
      y = c(actual, proj)
      #type <- c(c(rep("actual",length(selectionScurves[,i]))), c(rep("projection", length(caseFit) - length(selectionScurves[,i]))))
      line <- data.frame(x,y, c(rep("Actual", length(actual)), rep("Projected", length(proj))))
      colnames(line) <- c("Reported_Date", "n", "type")
      caseSelectionLines[[i]] <- list("line"=line, "names" = selectionNames[[i-1]], "color"=selectionColors[i-1])
  }

  #build the line for  "the rest"
  x <- caseCountData$Reported_Date
  actual = caseCountFitLine$n[1:length(selectionScurves[,1])] * selectionScurves[,1]
  actual <- actual[!is.na(actual)]
  if (length(actual) == length(caseCountData$Reported_Date)){
    proj = NULL
  } else{
    proj = caseCountFitLine$n[(length(selectionScurves[,i])+1):length(caseCountFitLine$n)] * selectionScurvesExtended[,1][(length(selectionScurves[,1])+1):length(caseCountFitLine$n)]
    proj <- proj[!is.na(proj)]
  }
  y = unique(c(actual, proj))
  line <- data.frame(x,y, c(rep("Actual", length(actual)), rep("Projected", length(proj))))
  colnames(line) <- c("Reported_Date", "n", "type")
  caseSelectionLines[[ncol(selectionScurves) + 1]] <- list("line"=line, "names" = "The Rest", "color"="black")
  
  #gather the population data for /100000 individual normalization
  if (!selectionEstimateObject$region %in% c("Alberta", "Ontario", "Quebec")){
    population <- as.numeric(populationData[2,selectionEstimateObject$region]* 7)
  } else {
    population <- (as.numeric(populationData[2,selectionEstimateObject$region]) )
  }
  #pass the lines into the plotting function
  plotCaseCountByDate2(caseCountData, rev(caseSelectionLines),population,region=selectionEstimateObject$region, maxdate = maxdate, order=mutantNames, saveToFile=saveToFile)
}

```


```{r, warning=FALSE, message=FALSE, fig.height=8, fig.width=11, warning=FALSE, out.width="100%", results='asis', echo=FALSE}
#plots the case count selection plots with each pronvince as a tab.
apply(all.regions,1,function(reg){
  cat("###", reg[["shortname"]], "\n")
  cat("####",reg[["name"]],"\n","\n")
  if (is.null(selectionEstimateFits[[reg[["shortname"]]]]$plot1)){
    print(getEmptyErrorPlotWithMessage("Not enough genomic data available."))
  } else if (reg[["shortname"]] %in% names(caseData)){
      p<-plotCaseCountSelection(caseData[[reg[["shortname"]]]], selectionEstimateFits[[reg[["shortname"]]]])#,reg[["shortname"]])
      print(p)
  } else {
      print(getEmptyErrorPlotWithMessage("No case count data available."))
  }
  cat("\n\n")})
  
```

### {-}

<hr style="border:1px solid gray">

# Fastest growing lineages {.tabset .tabset-fade}

Here we show the selection estimates and their 95% confidence intervals for Pango lineages with more than 10 sequences in Canada since `r startdate` and with enough data to estimate the confidence interval. Each selection estimate measures the growth rate relative to `r individualSelectionPlotReference[[1]]` stricto (i.e., sequences designated as `r individualSelectionPlotReference[[1]]` and not its descendants, like `r individualSelectionPlotReference[[1]]`.1). Plots showing the change in variant frequency over time in Canada are given below for lineages with more than 50 sequences.

```{r getCoefficientTable, echo = F, warning=FALSE, message=FALSE}
source("scripts/plot_growing_lineages.R")

paramselected=allparams[sapply(allparams,function(p){if(!any(is.na(p$fit))){
x=p$mut[[1]];
(p$fit)$fit[["s1"]]>0 & 
sum(p$toplot$n2)>10
}else{FALSE}})]

coefficientTable <- plot_growing_lineage(paramselected,makeplot=FALSE)

```

## Plot (stricto) {.tabset .tabset-fade}

Growth advantage of 0-5% corresponds to doubling times of more than two weeks, with 5-10% reflecting one to two week doubling times and over 10% representing significant growth of less than one week doubling time. Note that estimating selection of sub-variants with low sequence counts (less than 100 dots) is prone to error, such as mistaking one-time super spreader events or pulses of sequence data from one region as selection. Estimates with lower sequence counts in one region should be considered as very preliminary. For Canada-wide plot, a dot with a circle border indicates lineages with a positive selection coefficient in multiple provinces. 


```{r buildStrictoPlots, results='asis', echo = F, warning=FALSE, fig.height=7}


selectparam <- function(p,reg){
  if(!any(is.na(p$fit))){
    x=p$mut[[1]];
    p$region==reg&
    (p$fit)$fit[["s1"]]>0 & 
    sum(p$toplot$n2)>10&
    substr(x, nchar(x), nchar(x))!="*"
  }else{FALSE}
  }


for (i in 1:length(all.regions[["name"]])) {
  cat("###", all.regions[['shortname']][[i]], "\n")
  if (all.regions[['name']][[i]] == "Canada"){
      cat("#### Plot single lineages in", all.regions[['name']][[i]],"*\n","\n")
  }else{
      cat("#### Plot single lineages in", all.regions[['name']][[i]],"\n","\n")
  }
  paramselected=allparams[sapply(allparams,function(x){selectparam(x,all.regions[['name']][[i]])})]
  n=min(25,length(paramselected))
  if(n!=0){
    #print(plot_growing_lineage(paramselected[1:n]))
    p<-plot_growing_lineage(paramselected[1:n], coefficientTable=coefficientTable) #A%>% as_widget()
    #print(htmltools::tagList(ggplotly(p, height = 700)))
    print(p)
  }else{
    p<-(getEmptyErrorPlotWithMessage("Not enough data available.") )
    print(p)
  }
  cat("\n\n")
}
```

## Plot (non stricto)  {.tabset .tabset-fade}
      
This plot highlights the groups of related lineages that are growing fastest (e.g., BQ.1* is the monophyletic clade that includes BQ.1.1 and all other BQ.1 sublineages. 

Again, Growth advantage of 0-5% corresponds to doubling times of more than two weeks, with 5-10% reflecting one to two week doubling times and over 10% representing significant growth of less than one week doubling time. Note that estimating selection of sub-variants with low sequence counts (less than 100 dots) is prone to error, such as mistaking one-time super spreader events or pulses of sequence data from one region as selection. Estimates with lower sequence counts in one region should be considered as very preliminary. For Canada-wide plot, a dot with a circle border indicates lineages with a positive selection coefficient in multiple provinces. 


```{r buildNonStrictoPlots, results='asis', echo=FALSE, warning=FALSE, fig.height=7}
source("scripts/plot_growing_lineages.R")
selectparam <- function(p,reg){
  if(!any(is.na(p$fit))){
    x=p$mut[[1]];
    p$region==reg&
    (p$fit)$fit[["s1"]]>0 & 
    sum(p$toplot$n2)>10&
    substr(x, nchar(x), nchar(x))=="*"
  }else{FALSE}
  }

for (i in 1:length(all.regions[["name"]])) {
  cat("###", all.regions[['shortname']][[i]], "\n")
  cat("#### Plot single lineages in", all.regions[['name']][[i]],"\n","\n")
  paramselected=allparams[sapply(allparams,function(x){selectparam(x,all.regions[['name']][[i]])})]
  n=min(25,length(paramselected))
  if(n!=0){
    #print(plot_growing_lineage(paramselected[1:n]))
    p<-plot_growing_lineage(paramselected[1:n], coefficientTable=coefficientTable) #A%>% as_widget()
    #print(htmltools::tagList(ggplotly(p, height = 700)))
    print(p)
  }else{
    p<-(getEmptyErrorPlotWithMessage("Not enough data available.") )
    print(p)
  }
  cat("\n\n")
}

```

```{r overview of the selection estimator table, echo=FALSE}
dataDownload <- coefficientTable %>% mutate(sel_coeff = sel_coeff / 100)  %>% mutate(low_CI = low_CI / 100)  %>% mutate(high_CI = high_CI / 100)
link = DisplayHTMLTableDownloadLink(dataDownload, "GrowingLineages")
#DisplayHTMLTable(plot_growing_lineage(paramselected,makeplot=FALSE))
```

## Table of all the selection estimates 

<a href="./downloads/GrowingLineages.tsv" target="_blank" class="btn btn-primary">Download Table</a>

## {-}

<hr style="border:1px solid gray">

# Sublineages selection 

## BA.2 and XBB sublineages {.tabset .tabset-fade}

Here we show the trends of the various BA.2.\* sublineages over time, relative to the frequency of `r individualSelectionPlotReference[[1]]` by itself (shown for sublineages with at least 50 (Canada) or 20 (provinces) cases). Proportions shown here are only among `r individualSelectionPlotReference[[1]]` (stricto) and the lineage illustrated. Note that these plots are not necessarily representative of trends in each province and that mixing of data from different provinces may lead to shifts in frequency that are not due to selection.

```{r general settings for selection estimator BA.2, warning=FALSE, echo=FALSE, message=FALSE}
source("scripts/scanlineages.R")
source("scripts/plot_selection_estimator.R")

#function to plot all the indivudal variant's selection plot
plotIndividualSublineageSelection <- function(region, minimumSampleCount, parentalNodes, maxdate){
  allSublineagePlots <- list()
  plotCounter <- 0
  #loop through all params and find...
  for(i in seq(1:length(allparams))) {
    if(!any(is.na(allparams[[i]]$fit)) && # fit not NA
       allparams[[i]]$region==region &&  #region is the same as what we want
       substr(allparams[[i]]$mut[[1]], nchar(allparams[[i]]$mut[[1]]), nchar(allparams[[i]]$mut[[1]]))!="*" && #is the same lineage
       isSubLineage(parentalNodes,allparams[[i]]$mut[[1]])){ #is child lineage of parentalNodes 
      if((allparams[[i]]$fit)$fit[["s1"]]>0 && sum(allparams[[i]]$toplot$n2)>minimumSampleCount){ #the numebr of samples are greater than the minimum we want.
          plotCounter <- plotCounter + 1
          #plot all the variants and save them to a named list
          allSublineagePlots[[plotCounter]] <- list("variant" = allparams[[i]]$mut[[1]], "plot" = plotIndividualSelectionPlots.ggplot(plotparam=allparams[[i]], maxdate = maxdate))
      }
    }
  }
  return(allSublineagePlots)
}

#generate  selection variant plots for BA2 and BA5 variants for each province
individualSublineageSelectionPlots.BA2 <- list() 
individualSublineageSelectionPlots.BA5 <- list() 
for (i in 1:length(all.regions$name)){
  minCount = ifelse(all.regions$shortname[[i]] == "Canada", 50, 20)
  individualSublineageSelectionPlots.BA2[[all.regions$shortname[[i]]]] = plotIndividualSublineageSelection(region=all.regions$name[[i]], minCount, "BA.2*", maxdate)
  individualSublineageSelectionPlots.BA5[[all.regions$shortname[[i]]]] = plotIndividualSublineageSelection(region=all.regions$name[[i]], minCount, "BA.5*", maxdate)
}


```

```{r, warning=FALSE, message=FALSE, echo = FALSE, fig.height=5, fig.width=5, warning=FALSE, out.width="33%", results='asis', dpi=150}
#prints the BA.2 plots
apply(all.regions,1,function(reg){
  cat("###", reg[["shortname"]], "\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("####",reg[["name"]],"\n","\n")
  if (length(individualSublineageSelectionPlots.BA2[[reg[["shortname"]]]])>0){
      i = 0
      for (plot in individualSublineageSelectionPlots.BA2[[reg[["shortname"]]]]){
        i = i + 1
        if (i==4){
          cat ("<P>")
          cat("<details>")
          cat ("<summary><b>Only the three most strongly selected variants are displayed. Click here to see the rest.</b></summary>")
        }
        print(plot$plot)
      }
      if (i>3){cat("</details>")}
  }else{
    print(getEmptyErrorPlotWithMessage("Not enough data avilable"))
  }
  cat("\n\n")})

```

## BA.5 sublineages {.tabset .tabset-fade}

Here we show the trends of the various BA.5* sublineages over time, relative to the frequency of `r individualSelectionPlotReference[[1]]` by itself (shown for sublineages with at least 50 (Canada) or 20 (provinces) cases). Proportions shown here are only among `r individualSelectionPlotReference[[1]]` (stricto) and the lineage illustrated. Note that these plots are not necessarily representative of trends in each province and that mixing of data from different provinces may lead to shifts in frequency that are not due to selection.

```{r, warning=FALSE, message=FALSE, echo = FALSE, fig.height=5, fig.width=5, warning=FALSE, out.width="33%", results='asis', dpi=150}
#prints the BA.5 Plots
apply(all.regions,1,function(reg){
  cat("###", reg[["shortname"]], "\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("####",reg[["name"]],"\n","\n")
  if (length(individualSublineageSelectionPlots.BA5[[reg[["shortname"]]]])>0){
      i = 0
      for (plot in individualSublineageSelectionPlots.BA5[[reg[["shortname"]]]]){
        i = i + 1
        if (i==4){
          cat ("")
          cat("<details>")
          cat ("<summary><b>Only the variants with the 3 highest selection coefficients are displayed. Click here to see the rest.</b></summary>")
        }
        print(plot$plot)
      }
      if (i>3){cat("</details>")}
  }else{
    print(getEmptyErrorPlotWithMessage("Not enough data avilable"))
  }
  cat("\n\n")})

```

## {-}
