Here we examine the relative rate of spread of the different sublineages of Omicron currently in Canada. Specifically, we determine if a new or emerging lineage has a selective advantage, s (and by how much), against a reference lineage previously common in Canada (see the methods for more details about selection and how it is estimated).

Currently, the major group of Omicron lineages rising in frequency are the XBB.\* group with BQ.*,which descend from BA.5\*, nearing a peak. We first show this growth of XBB\* and other BQ.\*, relative to the remaining strains, which consist predominantly of BA.5\* (excluding BQ.\*). Left plot: y-axis is the proportion of sub-lineages BQ.\* and XBB.\* relative to the remaining strains; right plot: y-axis describes the logit function, log(freq(XBB.\* or BQ\*)/freq(the rest)), which gives a straight line whose slope is the selection coefficient if selection is constant over time (see methods).

For comparison, Alpha had a selective advantage of s ~ 6%-11% per day over preexisting SARS-CoV-2 lineages, and Delta had a selective advantage of about 10% per day over Alpha.

**Caveat:**
These selection analyses must be interpreted with caution due to the potential for non-representative sampling, lags in reporting, and spatial heterogeneity in prevalence of different sublineages across Canada. Provinces that do not have at least 20 sequences of XBB.\*, BQ.\*, and other lineages during this time frame are not displayed.

```{r, warning=FALSE, message=FALSE}
source("scripts/plot_selection_estimator.R")

startpar <- list(p=c(0.2, 0.05), s=c(0.05, 0.1))
#startpar <- list(p=c(0.5, 0.1, 0.01), s=c(0, 0.05, 0.05))

setAll=getAllStrictoLineages(meta)
#sublineages_BA5 <- getStrictoSubLineages("BA.5*",meta)
sublineages_BQ <- getStrictoSubLineages("BQ*",meta)
sublineages_XBB <- getStrictoSubLineages("XBB*",meta)

#setAll=setAll[!setAll %in% sublineages_BA5]
setAll=setAll[!setAll %in% sublineages_BQ]
setAll=setAll[!setAll %in% sublineages_XBB]

mutants = list(sublineages_BQ, sublineages_XBB)
mutantNames = list("BQ*", "XBB*", "the rest") #"
col <- c(pal["Omicron BQ"], "XBB"="lightblue")

#mutants = list(sublineages_BA5, sublineages_BQ, sublineages_XBB)
#mutantNames = list("BA.5*","BQ*", "XBB*", "the rest") #"
#color for each lineage
#col <- c(pal["Omicron BA.5"], pal["Omicron BQ"], "XBB"="lightblue")

#Set a starting date
#Note that the startdate shouldn't be too much before both alleles become common
#or rare migration events that die off could throw off the estimation procedure 
#(so that the parameter estimates account for the presence of those alleles long in the past).
startdate<-max(meta$sample_collection_date)-days(120) #Using a later date with less sampling noise

sub.plot.selection.estimate <- function(region,maxdate=NA){
  obj=plot.selection.estimate.ggplot(region=region,
                                        startdate=startdate, startpar=startpar,
                                        reference=c(setAll),
                                        mutants=mutants,
                                        names=mutantNames,
                                        maxdate=maxdate, col=col)
  return(obj)
}

selectionEstimateFits <- list()
selectionEstimateFits[["Canada"]] = sub.plot.selection.estimate(region="Canada")
maxdate=selectionEstimateFits[["Canada"]]$date
selectionEstimateFits[["BC"]] <- sub.plot.selection.estimate(region="Alberta",maxdate=maxdate) #temp hax to get this to build
selectionEstimateFits[["BC"]]$plot1 <- NULL
#sub.plot.selection.estimate(region="British Columbia",maxdate=maxdate)
selectionEstimateFits[["AB"]] <- sub.plot.selection.estimate(region="Alberta",maxdate=maxdate)
selectionEstimateFits[["SK"]] <- sub.plot.selection.estimate(region="Alberta",maxdate=maxdate) #temp hax to get this to build
selectionEstimateFits[["SK"]]$plot1 <- NULL

selectionEstimateFits[["MB"]] <- sub.plot.selection.estimate(region="Manitoba",maxdate=maxdate)
selectionEstimateFits[["ON"]] <- sub.plot.selection.estimate(region="Ontario",maxdate=maxdate)
selectionEstimateFits[["QC"]] <- sub.plot.selection.estimate(region="Quebec",maxdate=maxdate)
selectionEstimateFits[["East"]] <- sub.plot.selection.estimate(region="East provinces (NL+NS+NB)",maxdate=maxdate)
#####################################################
# tabs for displaying in notebook
#each PT tab should have curve plot and breakpoint plot side by side
```

```{r, warning=FALSE, message=FALSE, fig.height=8, fig.width5, warning=FALSE, out.width="50%", results='asis'}
dataAvailable.regions <- all.regions %>% filter(shortname %in% c("Canada", "BC", "AB", "SK", "MA", "ON", "QC", "East"))
apply(dataAvailable.regions,1,function(reg){
  cat("###", reg[["shortname"]], "\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("####",reg[["name"]],"\n","\n")
  if (is.null(selectionEstimateFits[[reg[["shortname"]]]]$plot1))
  {cat("Not enough data available")}
  else {
    print(selectionEstimateFits[[reg[["shortname"]]]]$plot1)
    print(selectionEstimateFits[[reg[["shortname"]]]]$plot2)
  }
  cat("\n\n")})
  
```

## {.unlisted .unnumbered}

## Selection by Case Count {.tabset}

```{r case count selection estimator setup}
source("scripts/plot_case_count_selection_estimator.R")
caseData <- parseCaseData(maxdate, params$datadir)
populationData <- read.table("resources/2022Population.tsv", header=T, sep='\t', check.names = F)
plotCaseCountSelection <- function(caseCountData, selectionEstimateObject, filename=NA){
  #caseCountData <- caseData$Canada
  #selectionEstimateObject<- selectionEstimateFits$Canada
  #filename=NA
  if (selectionEstimateObject$region == "Canada"){dayToCut = 2} else { dayToCut = 5 }
  caseCountData$report_type <- "Accurate"
  caseCountData$report_type[(nrow(caseCountData)-dayToCut+1):nrow(caseCountData)] <- "UnderReported"
  caseFitModel <- getCaseCountSmoothFitWithLambda(caseCountData %>% filter(report_type=="Accurate"))
  caseFitLinePredictedValues <- c((10^(predict(caseFitModel, data.frame(x=(nrow(caseCountData)-dayToCut+1):nrow(caseCountData)))[[2]]))$x)
  caseFitLineValues <- 10^(caseFitModel$y)
  caseCountFitLine<-data.frame(caseCountData$Reported_Date, c(caseFitLineValues, c(caseFitLinePredictedValues)), caseCountData$report_type)
  colnames(caseCountFitLine) <- c("Reported_Date", "n", "type")
  caseCountFitLine <- caseCountFitLine %>% mutate(type = ifelse(type=="UnderReported", "Projected", "Actual"))
  
  selectionFit <- selectionEstimateObject$fit
  selectionScurves <- selectionEstimateObject$scurves
  selectionScurvesExtended <- selectionEstimateObject$scurvesExtended
  if (selectionEstimateObject$region == "Canada"){
      selectionScurves <- selectionEstimateObject$scurves[seq(1, nrow(selectionEstimateObject$scurves), 7), ]
      selectionScurvesExtended <- selectionEstimateObject$scurvesExtended[seq(1, nrow(selectionEstimateObject$scurvesExtended), 7), ]
  }
  selectionNames <- selectionEstimateObject$names
  selectionColors <- selectionEstimateObject$color
  caseSelectionLines <- list()
  #build to total case line.
  caseSelectionLines[[1]] <- list("line"=caseCountFitLine, "names" = "CaseCount", "color"="limegreen")

  #build line for each variant
  for (i in seq(from=2, to=ncol(selectionScurves))){
      x = caseCountData$Reported_Date
      actual = caseCountFitLine$n[1:length(selectionScurves[,i])] * selectionScurves[,i]
      actual <- actual[!is.na(actual)]
      if (length(actual) == length(caseCountData$Reported_Date)){
        proj = NULL
      } else{
        proj = caseCountFitLine$n[(length(selectionScurves[,i])+1):length(caseCountFitLine$n)] *   selectionScurvesExtended[,i][(length(selectionScurves[,i])+1):length(caseCountFitLine$n)]
        proj <- proj[!is.na(proj)]
      }
      y = c(actual, proj)
      #type <- c(c(rep("actual",length(selectionScurves[,i]))), c(rep("projection", length(caseFit) - length(selectionScurves[,i]))))
      line <- data.frame(x,y, c(rep("Actual", length(actual)), rep("Projected", length(proj))))
      colnames(line) <- c("Reported_Date", "n", "type")
      caseSelectionLines[[i]] <- list("line"=line, "names" = selectionNames[[i-1]], "color"=selectionColors[i-1])
  }

  #build the line for  "the rest"
  x <- caseCountData$Reported_Date
  actual = caseCountFitLine$n[1:length(selectionScurves[,1])] * selectionScurves[,1]
  actual <- actual[!is.na(actual)]
  if (length(actual) == length(caseCountData$Reported_Date)){
    proj = NULL
  } else{
    proj = caseCountFitLine$n[(length(selectionScurves[,i])+1):length(caseCountFitLine$n)] * selectionScurvesExtended[,1][(length(selectionScurves[,1])+1):length(caseCountFitLine$n)]
    proj <- proj[!is.na(proj)]
  }
  y = unique(c(actual, proj))
  line <- data.frame(x,y, c(rep("Actual", length(actual)), rep("Projected", length(proj))))
  colnames(line) <- c("Reported_Date", "n", "type")
  caseSelectionLines[[ncol(selectionScurves) + 1]] <- list("line"=line, "names" = "The Rest", "color"="black")
  
  if (selectionEstimateObject$region == "Canada"){
    population <- as.numeric(populationData[2,selectionEstimateObject$region]* 7)
  } 
  else {
    population <- (as.numeric(populationData[2,selectionEstimateObject$region]) )
  }
  plotCaseCountByDate2(caseCountData, rev(caseSelectionLines),population,filename=filename )
}

```

These plots track the reported cases among people aged over 70 per 100,000 individuals (green dots). This age group is more reliably tested in Canada and is thus used to describe overall trends in COVID-19. A cubic spline is fit to the log of case counts (top curve), with the recent trends used to estimate the daily exponential growth rate r in COVID-19 cases on the last day of accurate case counts.  The fit from the “Selection on Omicron” section above is used to show how each of the sub-lineages is growing or shrinking, with the corresponding exponential growth rate. The most recent case counts are generally underreported and are not used in the fits (dropping two weeks of data from Canada and five days of data from the provinces). The lambda value used for the spline fit is 0.001.

```{r, warning=FALSE, message=FALSE, fig.height=8, fig.width=11, warning=FALSE, out.width="100%", results='asis'}
dataAvailable.regions <- all.regions %>% filter(shortname %in% c("Canada", "BC", "AB", "SK", "MA", "ON", "QC", "East"))
apply(dataAvailable.regions,1,function(reg){
  cat("###", reg[["shortname"]], "\n")
  cat("####",reg[["name"]],"\n","\n")
  if (is.null(selectionEstimateFits[[reg[["shortname"]]]]$plot1)){
    cat("Not enough genomic data available.")
  } else if (reg[["shortname"]] %in% names(caseData)){
      p<-plotCaseCountSelection(caseData[[reg[["shortname"]]]], selectionEstimateFits[[reg[["shortname"]]]])#,reg[["shortname"]])
      print(p)
  } else {
    cat("No case count data available.")
  }

  cat("\n\n")})
  
```


## {.unlisted .unnumbered}

### BA.2 sublineages selection {.tabset}

```{r general settings for selection estimator BA.2}
source("scripts/scanlineages.R")
plotsubvariant <- function(region,min,parentalnode){
  nplot=0
  for(p in allparams) {
    if(!any(is.na(p$fit)) && 
       p$region==region && 
       substr(p$mut[[1]], nchar(p$mut[[1]]), nchar(p$mut[[1]]))!="*" &&
       isSubLineage(parentalnode,p$mut[[1]])){
      if((p$fit)$fit[["s1"]]>0 && sum(p$toplot$n2)>min){
          plot.selection(plotparam=p)
          nplot=nplot+1
      }
    }
  }
  if(nplot==0){
    print("Not enough data available")
  }
}


plotba2 <- function(region,min){
  plotsubvariant(region,min,"BA.2*")
}



plotba5 <- function(region,min){
  plotsubvariant(region,min,"BA.5*")
}


plotothers <- function(region,min){
  nplot=0
  for(p in allparams) {
    if(!any(is.na(p$fit)) && 
       p$region==region && 
       substr(p$mut[[1]], nchar(p$mut[[1]]), nchar(p$mut[[1]]))!="*" &&
       !isSubLineage("BA.2*",p$mut[[1]])){
      if((p$fit)$fit[["s1"]]>0 && sum(p$toplot$n2)>min){
          plot.selection(plotparam=p)
          nplot=nplot+1
      }
    }
  }
  if(nplot==0){
    print("Not enough data available")
  }
}


```


Here we show the trends of the various BA.2.* sublineages over time, relative to the frequency of `r reference[[1]]` by itself (shown for sublineages with at least 50 (Canada) or 20 (provinces) cases). Proportions shown here are only among `r reference[[1]]` (stricto) and the lineage illustrated. Note that these plots are not necessarily representative of trends in each province and that mixing of data from different provinces may lead to  shifts in frequency that are not due to selection.


```{r, warning=FALSE, message=FALSE, fig.height=5, fig.width=5, warning=FALSE, out.width="33%", results='asis'}
tmp=apply(all.regions,1,function(reg){
  cat("####", reg[["shortname"]], "\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("#####",reg[["name"]],"\n","\n")
  plotba2(reg[["name"]],n_min)
  cat("\n\n")})
  
```



### BA.5 sublineages selection {.tabset}

Here we show the trends of the various BA.5# sublineages over time, relative to the frequency of `r reference[[1]]` by itself (shown for sublineages with at least 50 (Canada) or 20 (provinces) cases). Proportions shown here are only among `r reference[[1]]` (stricto) and the lineage illustrated. Note that these plots are not necessarily representative of trends in each province and that mixing of data from different provinces may lead to  shifts in frequency that are not due to selection.



```{r, warning=FALSE, message=FALSE, fig.height=5, fig.width=5, warning=FALSE, out.width="33%", results='asis'}
tmp=apply(all.regions,1,function(reg){
  cat("####", reg[["shortname"]], "\n")
  n_min=20
  if(reg[["name"]]=="Canada"){n_min=50}
  cat("#####",reg[["name"]],"(n>",n_min,")\n","\n")
  plotba5(reg[["name"]],n_min)
  cat("\n\n")})
  
```

<!-- ### Other sublineages selection {.tabset} -->

<!-- Here we show the trends of the various other sublineages over time, relative to the frequency of `r reference[[1]]` by itself (shown for sublineages with at least 50 (Canada) or 20 (provinces) cases). Proportions shown here are only among `r reference[[1]]` (stricto) and the lineage illustrated. Note that these plots are not necessarily representative of trends in each province and that mixing of data from different provinces may lead to  shifts in frequency that are not due to selection. -->



<!-- ```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5, fig.show="hold", fig.width=5, warning=FALSE, out.width="33%", results='asis'} -->
<!-- tmp=apply(all.regions,1,function(reg){ -->
<!--   cat("####", reg[["shortname"]], "\n") -->
<!--   n_min=20 -->
<!--   if(reg[["name"]]=="Canada"){n_min=50} -->
<!--   cat("#####",reg[["name"]],"(n>",n_min,")\n","\n") -->
<!--   plotothers(reg[["name"]],n_min) -->
<!--   cat("\n\n")}) -->

<!-- ``` -->
